### ISSUE TITLE
Docs reorganisation (again!)

### ISSUE THREAD

**Issue by Matthew-Jennings** [pymedphys/pymedphys#1665]
Hi all,

I think the docs could really benefit from some simplification. A while ago, a decision was made to embrace a framework that included _background_, _tutorials_, _how-to-guides_ and _reference_. In retrosprect, I think this was premature and I think it's partially responsible for the docs feeling a little skeleton-y. The sidebar doesn't nicely match this framework.

I'd like to brainstorm opportunities for simplification here. One option for consideration is:

- Separate the docs very clearly (including the sidebar) into:
   - Overview
      - Discuss major features/modules (e.g. gamma, anaonymisation, etc.) of PyMedPhys and include links to user/contributors guides
         - _this enables a reader to find what they need according to feature, which is one natural path a reader is likely to want to take_
   - User guide
      - Library
      - App
      - CLI
   - Contributors guide
      - _TBD_

- Combine _tutorials_ and _how-tos_. I don't think it is useful to separate these until there is a marked different between them. The introduce clutter is not worth the added categorisation.

- Explore nesting of sidebar to improve organisation. If nesting is not possible, simply accept that not all pages need to be linked to sidebar.

-------------------------

Suggestions/thoughts? If there is enough support, I can make a PR work a live draft. But, although I have my own opinions, I'm very keen to hear from other PyMedPhys contributors/users - especially new ones.

**Comment by SimonBiggs** [comment]
Please make sure before anyone comments, if we could make sure we have all read through the complete justification and approach detailed at:

https://diataxis.fr/

(It would also help us align better to the approach)

**Comment by SimonBiggs** [comment]
> Separate the docs very clearly (including the sidebar) into:
> 
> * Overview
>   
>   * Discuss major features/modules (e.g. gamma, anaonymisation, etc.) of PyMedPhys and include links to user/contributors guides
>     
>     * _this enables a reader to find what they need according to feature, which is one natural path a reader is likely to want to take_
> * User guide
>   
>   * Library
>   * App
>   * CLI
> * Contributors guide
>   
>   * Setups
>   * Contribution workflow
>   * Maintenance guide

I really like this break down, and it is much cleaner than what we have currently.

**Comment by Matthew-Jennings** [comment]
> Please make sure before anyone comments, if we could make sure we have all read through the complete justification and approach detailed at:
> 
> https://diataxis.fr/
> 
> (It would also help us align better to the approach)

I guess I'd just add that the diataxis framework could still be applied, just not at top-level.

Though I would still argue that it is an over-complicated framework for _our_ needs and at our current level of documentation. We could consider re-embracing as our documentation grows.

A probably stronger argument for my proposed simplification is centred around _familiarity_. [Numpy](https://numpy.org/doc/stable/), [Scipy](https://docs.scipy.org/doc/scipy/) & [Astropy](https://docs.astropy.org/en/stable/index.html)'s docs appear much closer to what I've detailed above and are more intuitively structured in my humble opinion.

**Comment by SimonBiggs** [comment]
> [Numpy](https://numpy.org/doc/stable/), [Scipy](https://docs.scipy.org/doc/scipy/) & [Astropy](https://docs.astropy.org/en/stable/index.html)'s docs appear much closer to what I've detailed above and are more intuitively structured in my humble opinion.

Most definitely. In fact, they go into exactly this approach you detail:

> I guess I'd just add that the diataxis framework could still be applied, just not at top-level.

![image](https://user-images.githubusercontent.com/6559099/173259423-489e22d2-7339-4fed-9ee7-c7a798d62b19.png)

Probably a key document for us to read through is the following:

https://diataxis.fr/tutorials-how-to/

Which explicitly details the difference between how-to and tutorials.

**Comment by SimonBiggs** [comment]
Also, here's numpy's justification for adopting https://diataxis.fr/:

https://labs.quansight.org/blog/2020/03/documentation-as-a-way-to-build-community/

**Comment by SimonBiggs** [comment]
Other big name open source groups have also done this quite well:

https://ubuntu.com/core/docs

**Comment by SimonBiggs** [comment]
And for another good example of the approach, from the person who created it:

https://docs.divio.com/en/latest/

**Comment by Matthew-Jennings** [comment]
One thing to consider: sometimes the best end-result framework is not the best interim one.

I suspect diataxis _is_ a lovely solution with well-fleshed out documentation. It just feels to me to be too early for that with the documentation as it currently stands. And I just don't see a strong, concerted effort from the PyMedPhys contributors to flesh out the docs to a degree that diataxis works well.

To me, the gaps in the documentation and it's "skeleton" feel are jarring. So, I'd prefer a structure that _evolves_ with the documentation.

It's almost a case of "form follows function", as it were. We don't _yet_ have "tutorials" that are meaningfully different from "how-to guides", despite what we've chosen to call them for now.

**Comment by Matthew-Jennings** [comment]
I should note though that it's the top-level structure and the added benefit of feature-categorised links in the overview or somewhere else useful that are the big things I'd like to see changed/added.

The diataxis stuff is a minor point to me.

**Comment by Matthew-Jennings** [comment]
> Might I suggest the following tweak:
> 
> * Overview
> * Users Guide
>   
>   * Getting started
>     
>     * Quick start guide
>     * Using the Application
>     * Using the CLI
>     * Using the Library
>   * How-to guides
>   * Technical reference
>   * Background information
> * Contributors Guide
>   
>   * Getting started
>     
>     * Your first PR
>     * How your PR will be reviewed
>     * Joining the team!
>   * Tips and tricks

That seems pretty reasonable to me.

**Comment by SimonBiggs** [comment]
> That seems pretty reasonable to me.

Please iterate further :slightly_smiling_face:

**Comment by SimonBiggs** [comment]
Justification for nesting Application, CLI and Library under getting started is that it means for 95% of visitors to our documentation (people just testing the water who may or may not ever go onto use PyMedPhys) everything they want will be in the one place. Under "getting started". And under getting started we will have 4 tutorials, one being the quick-start guide. The other three covering the three main ways to interact with PyMedPhys.

Someone dipping their toes into the world of PyMedPhys may reasonably try out all three and then in their day-to-day work now have a reasonable idea of how they can use PyMedPhys to help them, and to then maybe one day decide to come back and actually use it.

> And I just don't see a strong, concerted effort from the PyMedPhys contributors to flesh out the docs to a degree that diataxis works well.

I'd like to argue that docs have actually been my primary focus for PyMedPhys over the last year. Especially if you count the Discourse group (https://pymedphys.discourse.group/) and detailed issues (https://github.com/pymedphys/pymedphys/issues/1656) as a form of documentation. And when asked by @nlambriICH how best the PyMedPhys community could be served I flagged two key areas as demonstrated by the issue pinning:

![image](https://user-images.githubusercontent.com/6559099/173261949-92451740-04e4-41b6-80a9-a093a531191b.png)

**Comment by Matthew-Jennings** [comment]
> 'd like to argue that docs (+ community) have actually been my primary focus for PyMedPhys over the last year.

That kinda proves my point though, mate. A year later and still far from feeling complete (not a criticism, a recognition of the gargantuan task that it is - especially for a community of volunteers).

**Comment by Matthew-Jennings** [comment]
> Justification for nesting Application, CLI and Library **under** getting started (as opposed to having them be the top-level descriptors) is that it means for 95% of visitors to our documentation (people just testing the water who may or may not ever go on to use PyMedPhys) everything they want will be in the one place. Under "getting started". And under getting started we will have 4 tutorials, one being the quick-start guide. The other three covering the three main ways to interact with PyMedPhys.
> 
> Someone dipping their toes into the world of PyMedPhys may reasonably try out all three and then in their day-to-day work now have a reasonable idea of how they can use PyMedPhys to help them, and to then maybe one day decide to come back and actually use it.

Sounds good to me.

**Comment by Matthew-Jennings** [comment]
> > A year later and still far from feeling complete
> 
> And I'd argue the biggest gaping hole is actually "tutorials". Those documents that 95% of PyMedPhys docs visitors actually need... (hence my strong response against the idea of doing away with tutorials for now).
> 
> But I think, it makes more sense to instead of calling them tutorials, call it a "fleshed out and complete getting started guide" that allows new users to dip their toes into all of the key parts of PyMedPhys.

Fair enough. It just seems to me better to have a tutorials section only once we actually have tutorials! Having one in the hope that some will be written soon is putting form over function and not my preferred approach.

That said, I'm not _strongly_ opposed to going for your "getting started" solution, especially if you're confident these getting started guides will be fleshed-out soon. So, I'll let the point rest unless/until third opinions enter the discussion.

**Comment by Matthew-Jennings** [comment]
> especially if you're confident these getting started guides will be fleshed out soon

On that point, I think I can help achieve this.

**Comment by SimonBiggs** [comment]
By the way, for some stats to backup my statement. Across the last three months visitors to docs.pymedphys.com we had 2254 new users and 2339 users (96% new users):

![image](https://user-images.githubusercontent.com/6559099/173263370-6b24fa2f-0aef-44d3-b1fd-1f62742f6cf8.png)

For an idea on which parts of the docs people are using most:

![image](https://user-images.githubusercontent.com/6559099/173263559-318ecb7a-db9c-420e-b130-e19856177f89.png)

Full report download:

[Analytics All Web Site Data Pages 20220301-20220612.csv](https://github.com/pymedphys/pymedphys/files/8887226/Analytics.All.Web.Site.Data.Pages.20220301-20220612.csv)

**Comment by Matthew-Jennings** [comment]
> some stats to backup my statement

which statement, sorry?

**Comment by SimonBiggs** [comment]
> > some stats to backup my statement
> 
> which statement, sorry?

That my 95% statement wasn't just pulled out of thin air (https://github.com/pymedphys/pymedphys/issues/1665#issuecomment-1153350663) -- it was actually pulled out of thin air... so I went looking for the stats ... ... and 96% vs 95% is a pretty good guess :slightly_smiling_face: 

> Those documents that 95% of PyMedPhys docs visitors actually need... (hence my strong response against the idea of doing away with tutorials for now).

---

Also, when adjusting the structure of the docs, we need to be careful to, if possible, not change the web-address of various pages. The following sessions came from direct URL links from out there on the interwebs:

![image](https://user-images.githubusercontent.com/6559099/173263947-944b633f-516a-47eb-8777-8227f7734cee.png)

So, if we change those addresses, we effectively break those links / user bookmarks / etc.

**Comment by Matthew-Jennings** [comment]
> > > some stats to backup my statement
> > 
> > 
> > which statement, sorry?
> 
> That my 95% statement wasn't just pulled out of thin air ([#1665 (comment)](https://github.com/pymedphys/pymedphys/issues/1665#issuecomment-1153350663)) -- it was actually pulled out of thin air... so I went looking for the stats ... ... and 96% vs 95% is a pretty good guess 🙂
> 
> > Those documents that 95% of PyMedPhys docs visitors actually need... (hence my strong response against the idea of doing away with tutorials for now).
> 
> Also, when adjusting the structure of the docs, we need to be careful to, if possible, not change the web-address of various pages. The following sessions came from direct URL links from out there on the interwebs:
> 
> ![image](https://user-images.githubusercontent.com/6559099/173263947-944b633f-516a-47eb-8777-8227f7734cee.png)
> 
> So, if we change those addresses, we effectively break those links / user bookmarks / etc.

Argh... so much for beta! :D

>Those documents that 95% of PyMedPhys docs visitors actually need... (hence my strong response against the idea of doing away with tutorials for now).

Oh yes, sorry, I do agree with this (even in the absence of the stats, I would've guessed this - perhaps with less accuracy though!). 

If we're going to make/continue with a concerted effort to add tutorials, it makes sense to have a spot for them. My disagreement lay in the idea of over-expanding the structure before having sufficient content to fill it and for the content to remain lacking for an extended period.

I think we can settle on keeping with diataxis in the manner you've suggested and combine it with this concerted effort that you've already started and I and others can hopefully help boost.

**Comment by SimonBiggs** [comment]
> concerted effort that you've already started and I and others can hopefully help boost.

I think honour where honour is due. @nlambriICH has paved the way here :slightly_smiling_face:, I have just been a sign post. (And plan to continue to be a sign post for the foreseeable future).

**Comment by SimonBiggs** [comment]
> That sounds like a good solution

And... potentially we should collect a list of URLs which we put within a nighty scheduled test runner that verify those URLs aren't broken? I would be happy to set that up if you @Matthew-Jennings wanted to spear-head this documentation effort?

**Comment by Matthew-Jennings** [comment]
> I would be happy to set that up if you @Matthew-Jennings wanted to spear-head this documentation effort?

Yep, happy to!

**Comment by SimonBiggs** [comment]
> Argh... so much for beta! :D

...also... I think it's probably time for us to leave beta... (https://github.com/pymedphys/pymedphys/issues/1056)

**Comment by nlambriICH** [comment]
I agree that the documentation structure could be improved.

Probably the best approach for the new tutorials section would have been to put it into the official docs once some of them had been created (perharps by creating an issue with a task list of tutorials to be written).

Anyway, I think the proposed structure improves the current one, but: what do you think of the effort and ETA for such refactoring could be? It seems to me a pretty extensive work, as I don't know if we have all the proposed new sections already covered in the current docs.

Could this documentation refactoring be the step to transition to the 1.0 version of PyMedPhys?
Before that, are there any open issues or features that should be addressed first?

**Comment by SimonBiggs** [comment]
> Could this documentation refactoring be the step to transition to the 1.0 version of PyMedPhys?
> Before that, are there any open issues or features that should be addressed first?

A while back I detailed things that I thought needed to happen before a 1.0.0 release:

https://github.com/pymedphys/pymedphys/issues/1056

If I had the choice of a single focus though, I think the documentation is more important. And at the end of the day, I just think we need to treat PyMedPhys *as if it is* out of beta for a while, and once we're doing that successfully, that'll probably be a good time to take it out of beta.

**Comment by nlambriICH** [comment]
The proposed documentation structure looks good to me!

I could work on the PR section of the contributors guide (as I mentioned in #1359).

**Comment by SimonBiggs** [comment]
And having that top-level contributor document sorted out is a key requirement to be able to become part of NumFOCUS:

https://numfocus.org/projects-overview#:~:text=What%20are%20the%20requirements%20for%20a%20project%20to%20join%20NumFOCUS%3F

I suspect, any gaps within what it takes to be a NumFOCUS approved project is something we should be aspiring for, and it would be quite helpful to be able to come under the NumFOCUS banner.

**Comment by SimonBiggs** [comment]
As another flag, when clicking the following link within the documentation overview:

![image](https://user-images.githubusercontent.com/6559099/173452377-84b72c5d-3bae-48d4-9cb3-75d5285a10b7.png)

We're taken to the following page:

https://docs.pymedphys.com/lib/tutes/index.html

**Comment by Matthew-Jennings** [comment]
> Also @Matthew-Jennings, if you can think of ways to have more inter-linkage with the discourse mailing-list board let me know. Should I be bringing back the discourse comment section at the bottom of each page like we had before?
> 
> I have been trying to use the discourse board as a bit of an onramp for people to easily ask their questions etc and it has been having some success. But I suspect there may be benefit from having it be more prominent and accessible (just not sure how best to do it). Maybe an item within the side-bar navigation with the little "reference external" symbol to make it clear that the link is leaving the docs:
> 
> ![image](https://user-images.githubusercontent.com/6559099/173451846-337b5b75-2c18-4b3b-aebd-7040519412df.png)
> 
> I suspect under getting started we could have a link called "Mailing list (for questions and support)" or something like that? Hmm... maybe something as simple as "Get help" would be better?
> 
> * Users Guide
>   
>   * Getting started
>     
>     * Quick start guide
>     * Using the Application
>     * Using the CLI
>     * Using the Library
>     * **Getting help**

@SimonBiggs, if you really want to promote the Discourse, have an icon/link to it visible on every page (i.e. in the sidebar, or in some footer-like "frame), describe it in some detail on the PyMedPhys landing page (_with_ icon - visual cue), and refer to it in obvious places (like getting started docs).

I do think that the "plugin" we used to have was super ugly.

**Comment by SimonBiggs** [comment]
Just a heads up @Matthew-Jennings, we're getting this redirect pain fixed upstream instead of doing it all manually:

https://github.com/readthedocs/readthedocs.org/issues/9418#issuecomment-1232198041

**Comment by Matthew-Jennings** [comment]
Good news!

**Comment by TheTripleV** [comment]
> Just a heads up @Matthew-Jennings, this redirect pain is being fixed upstream (yayy 🙂):
> 
> [readthedocs/readthedocs.org#9418 (comment)](https://github.com/readthedocs/readthedocs.org/issues/9418#issuecomment-1232194730)

Hi, I dunno if you're in the same boat as me and need version level redirects or not but I've been using and maintaining https://github.com/wpilibsuite/sphinxext-rediraffe for a few years now to handle redirects manually. If you are just doing a reorg and moving files around, rediraffe does have a tool to use git to automatically generate redirects for moved files. Even if you don't want to use it for redirects and want to wait for RTD, the auto detection might be helpful as a one-time thing.

**Comment by SimonBiggs** [comment]
> Diff Checker
> The diff checker ensures that deleted/renamed files in your git repo are in your redirects.
> 
> To run the diff checker,
> 
> Set rediraffe_branch and rediraffe_redirects in conf.py.
> Run the rediraffecheckdiff builder.

@TheTripleV, that looks absolutely amazing!

@Matthew-Jennings, would you be willing to try and give this a spin over the new documentation reorg that you underwent?

Thank you, both of you! :slightly_smiling_face:
